### PR TITLE
Queries on all children

### PR DESCRIPTION
**Pull request by zudov**
I decided to try implementing something for #270

- added `queries` and `mkQueries`
- extended 'todo' example with "Mark all as completed" button (not the best showcase though as task list component has all the `TaskId`s in its state so it could as well just traverse over them)

### CODE REVIEW

**Comment by jdegoes** [comment]
I like the idea of this, but should it not return `Map p i` instead? You can always convert that to `List i`, but you cannot go the other way around.

**Comment by zudov** [comment]
@jdegoes `Map p i` makes sense. I didn't think about it.

**Comment by zudov** [comment]
It seems that the name of existing `query` function is a verb ("to query") rather than singular noun ("a query"). Probably the plural version of `query` should be something like `queryAll` rather than `queries`.

Any better ideas for naming?

**Comment by jdegoes** [comment]
`queryAll` seems like a good name.

**Comment by garyb** [comment]
:+1: for `queryAll` and the `Map` return value.

**Comment by zudov** [comment]
I think `mkQuery` and `queryAll` is a pretty self-contained addition and if that's good to go I can squash the commits so that it's ready to be merged.

There is still no way of querying all grandchild components, or querying all children of a top level component from the driver, but I don't think that such functionality is really needed. Also, adding something like that seem to require much more breaking changes to `ChildF` mechanism of pointing to the children.

File: src/Halogen/Component.purs [hunk 1]
```diff
@@ -58,6 +60,8 @@ import Halogen.Query.EventSource (EventSource(..), ParentEventSource(), runEvent
 import Halogen.Query.HalogenF (HalogenF(), HalogenFP(..), hoistHalogenF, transformHF)
 import Halogen.Query.StateF (StateF(..), mapState)
 
+import Unsafe.Coerce (unsafeCoerce)
```
Comment (garyb) [inline src/Halogen/Component.purs]: Wait, what now? :laughing:  (I assume this is leftover from hole driven development)

Comment (zudov) [inline src/Halogen/Component.purs]: Uups

Comment (zudov) [inline src/Halogen/Component.purs]: Fixed. thanks!

**Comment by zudov** [comment]
Hmm, just ended up in a situation where I have multiple child types and I want to query all children of specific type.

Need to add `mkQueries'` and `queryAll'`

**Comment by garyb** [comment]
Hah, yes. I did think about those cases, but I wasn't sure if the `ChildPath` prisms let you get out of it without yet more combinators... I guess it makes sense that that is not the case though, given we needed the `query'` / `mkQuery'` combinators.

**Comment by zudov** [comment]
@garyb It seem to type-check, I'll also check it with my code now.

**Comment by zudov** [comment]
I am getting some weird errors when using `queryAll'`, e.g. `queryAll cpR ...` works fine, but then if I stick `H.slot' cpL ...` in my dom this same `queryAll cpR` begins failing with "Error: Always Fails" (coming from Aff)

**Comment by zudov** [comment]
The latest commit makes a modification of multi-component example which shows the failure, pressing "Read states" buttons causes the "Error: Always fails", I digged it with debugger, but didn't find the solution. Any insights?

**Comment by zudov** [comment]
After further investigation, something like `query (Left FooSlot) (left (action FooAction))` works fine, but `queryAll (left (action FooAction))` fails. So it's not just about the primed functions.

**Comment by garyb** [comment]
- `"Always fails"` is what happens when `empty` is used in `Aff`.
- This happens when the `HaltHF` action from `HalogenF` is evaluated.
- This happens when you send a query with the "wrong value" to a child component.

So to `queryAll` you'll have to filter the children so as to only send queries to the children "on the left", in this case. Have a look at `Halogen.Component.transform` for a bit more on this.

**Comment by zudov** [comment]
@garyb Ha, that makes sense. I think I got to the second point by myself and was suspecting the third one.

I will try to implement it now. Thanks.

**Comment by garyb** [comment]
Yeah, it's a little unfortunate, I tried many approaches to make this fully safe using the type system, but I think we'd need dependant types to do so really.

The next best thing I could come up with was to say "as long as you use the `ChildPath` stuff with multi-typed child components it will always work", but as you've encountered now that's not totally airtight :wink:

**Comment by zudov** [comment]
@garyb Seems like I fixed the primed functions, now the return types are not confusing (`Map p i` instead of `Map p' i`), I should have noticed that just from types.

However, `mkQueries` and `queryAll` remain to be broken as I described previously. Looking at `mkQueries` context, it just seem to lack information needed to do such matching. Perhaps I miss something? As I see now, the only way to make `queryAll` not to fail is to require supplying a `ChildPath` which it would project with, and which would make it identical to `queryAll`.
```haskell
mkQueries
  :: forall s s' f f' p g i.
   . (Functor g, Ord p)
  => f' i
  -> QueryF s s' f f' g p (M.Map p i)
mkQueries ..
 where
  matchesQuery :: f' i -> p -> Boolean
  matchesQuery = ???
```

**Comment by zudov** [comment]
I also need to come up with more meaningful examples.

**Comment by garyb** [comment]
I'm not sure I understand how `mkQueries` and `queryAll` are broken, unless you mean when they're used in a situation where `mQueries'`/`queryAll'` should be used instead? I don't think there's a way to avoid that.

**Comment by zudov** [comment]
@garyb Yes, I meant the `queryAll (left (action FooAction))` thingy

**Comment by garyb** [comment]
Right... yeah, I think that's unfixable. We should just include a note in the documentation about those functions being unsuitable when there are mixed component types as children.

**Comment by zudov** [comment]
I refactored code a little and squashed commits, it now should be ready for more-or-less final review. Commit message provides a summary part of which could go to release notes.

**Comment by zudov** [comment]
Just wanted to ping here. Can we merge it, or we want to add something else? I've been using this patch for some time and it works great so far.

File: src/Halogen/Component.purs [hunk 2]
```diff
@@ -186,6 +190,33 @@ query'
...
+-- | Queries every child component that is currently installed. For use within
+-- | a parent component's `eval` or `peek` function.
+queryAll
+  :: forall s s' f f' g p i
+   . (Functor g, Ord p)
+  => f' i
+  -> Free (HalogenFP ParentEventSource s f (QueryF s s' f f' g p)) (M.Map p i)
+queryAll q = liftQuery (mkQueries q)
+
+childSlots
```
Comment (garyb) [inline src/Halogen/Component.purs]: Could you add a doc comment for this one?

Comment (zudov) [inline src/Halogen/Component.purs]: Yes, just a moment.

Comment (zudov) [inline src/Halogen/Component.purs]: Added "Returns slots of all currently installed child components."

**Comment by garyb** [comment]
Thanks!

**Comment by zudov** [comment]
@garyb @jdegoes Thanks for all help

**Pull request merged**